### PR TITLE
Pattern Detection in Declarative Deployment Models

### PR DESCRIPTION
**Pull request by awersching**
Enable pattern detection in topologies by extending and repurposing PRMs.

- Start and end date: 2020-10-28 to 2021-04-28
- Contributor: Adrian Wersching @awersching 
- Supervisor: Lukas Harzenetter @lharzenetter
- [ ] Ensure that you followed http://eclipse.github.io/winery/dev/ToolChain#github---prepare-pull-request.
- [x] Branch name checked. That means, the branch name starts with `thesis/`, `fix/`, ...
- [ ] Ensure that the commit message is [a good commit message](https://github.com/joelparkerhenderson/git_commit_message)
- [x] Ensure to use auto format in **all** files
- [x] Ensure that you appear in `NOTICE` at Copyright Holders
- [x] Tests created for changes
- [x] Documentation updated (if needed)
- [x] Screenshots added (for UI changes)

Screenshot_2021-03-11_20-12-03
behavior-pattern-mappings-table
grafic-mappings

### CODE REVIEW

File: org.eclipse.winery.model.adaptation/src/main/java/org/eclipse/winery/model/adaptation/substitution/refinement/AbstractRefinement.java [hunk 1]
```diff
@@ -38,13 +39,13 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public abstract class AbstractRefinement extends AbstractSubstitution {
+public abstract class AbstractRefinement extends AbstractSubstitution implements RefineServiceTemplate {
```
Comment (lharzenetter) [inline org.eclipse.winery.model.adaptation/src/main/java/org/eclipse/winery/model/adaptation/substitution/refinement/AbstractRefinement.java:42]: Why do you need this interface here?

File: org.eclipse.winery.model.adaptation/src/test/java/org/eclipse/winery/model/adaptation/substitution/patterndetection/BehaviorPatternDetectionTest.java [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+        behaviorPatternDetection.applyRefinement(refinementCandidate, topology);
+        assertEquals(topology.getNodeTemplateOrRelationshipTemplate().size(), 1);
+        policies = topology.getNodeTemplates().get(0).getPolicies().getPolicy();
+        assertEquals(policies.size(), 3);
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("noProp")));
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("oneProp")));
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("multiProps")));
+    }
+
+    private NamespaceManager namespaceManager() {
```
Comment (lharzenetter) [inline org.eclipse.winery.model.adaptation/src/test/java/org/eclipse/winery/model/adaptation/substitution/patterndetection/BehaviorPatternDetectionTest.java:301]: Isn't there a NamespaceManager for testing already?

File: org.eclipse.winery.model/org.eclipse.winery.model.tosca.canonical/src/main/java/org/eclipse/winery/model/tosca/extensions/OTBehaviorPatternMapping.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    public void accept(Visitor visitor) {
+        visitor.visit(this);
+    }
+
+    public static class Builder extends OTPrmMapping.Builder<Builder> {
+
+        private String behaviorPattern;
+        private OTPropertyKV property;
+
+        public Builder() {
```
Comment (lharzenetter) [inline org.eclipse.winery.model/org.eclipse.winery.model.tosca.canonical/src/main/java/org/eclipse/winery/model/tosca/extensions/OTBehaviorPatternMapping.java:83]: Do you need the default constuctor?

File: org.eclipse.winery.model/org.eclipse.winery.model.tosca.canonical/src/main/java/org/eclipse/winery/model/tosca/extensions/OTBehaviorPatternMapping.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            this.property = property;
+            return self();
+        }
+
+        public OTBehaviorPatternMapping build() {
+            return new OTBehaviorPatternMapping(this);
+        }
+
+        @Override
+        public @ADR(11) OTBehaviorPatternMapping.Builder self() {
```
Comment (lharzenetter) [inline org.eclipse.winery.model/org.eclipse.winery.model.tosca.canonical/src/main/java/org/eclipse/winery/model/tosca/extensions/OTBehaviorPatternMapping.java:106]: remove the ADR annotation

File: org.eclipse.winery.model/org.eclipse.winery.model.tosca.xml/pom.xml [hunk 5]
```diff
@@ -72,6 +72,12 @@
             <version>${org.slf4j}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.eclipse.winery</groupId>
+            <artifactId>org.eclipse.winery.model.tosca.canonical</artifactId>
```
Comment (lharzenetter) [inline org.eclipse.winery.model/org.eclipse.winery.model.tosca.xml/pom.xml:77]: Why do you add a dependency to the canonical model in the xml model? The XML model must not be aware of the wrapping model...

File: org.eclipse.winery.model/org.eclipse.winery.model.tosca.xml/src/main/java/org/eclipse/winery/model/tosca/xml/extensions/XOTBehaviorPatternMapping.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    public void accept(Visitor visitor) {
+        visitor.visit(this);
+    }
+
+    public static class Builder extends XOTPrmMapping.Builder<Builder> {
+
+        private String behaviorPattern;
+        private XOTPropertyKV property;
+
+        public Builder() {
```
Comment (lharzenetter) [inline org.eclipse.winery.model/org.eclipse.winery.model.tosca.xml/src/main/java/org/eclipse/winery/model/tosca/xml/extensions/XOTBehaviorPatternMapping.java:82]: Do you need the default constuctor?

File: org.eclipse.winery.model/org.eclipse.winery.model.tosca.xml/src/main/java/org/eclipse/winery/model/tosca/xml/extensions/XOTBehaviorPatternMapping.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            this.property = property;
+            return self();
+        }
+
+        public XOTBehaviorPatternMapping build() {
+            return new XOTBehaviorPatternMapping(this);
+        }
+
+        @Override
+        public @ADR(11) Builder self() {
```
Comment (lharzenetter) [inline org.eclipse.winery.model/org.eclipse.winery.model.tosca.xml/src/main/java/org/eclipse/winery/model/tosca/xml/extensions/XOTBehaviorPatternMapping.java:105]: Remove the ADR annotation

File: org.eclipse.winery.model/org.eclipse.winery.model.tosca.xml/src/main/java/org/eclipse/winery/model/tosca/xml/extensions/XOTPatternRefinementModel.java [hunk 8]
```diff
@@ -13,23 +13,81 @@
...
+        this.behaviorPatternMappings = behaviorPatternMappings;
     }
 
     public static class Builder extends XOTTopologyFragmentRefinementModel.Builder {
-        public Builder() { }
+
+        private boolean isPdrm;
+        private List<XOTBehaviorPatternMapping> behaviorPatternMappings;
+
+        public Builder() {
```
Comment (lharzenetter) [inline org.eclipse.winery.model/org.eclipse.winery.model.tosca.xml/src/main/java/org/eclipse/winery/model/tosca/xml/extensions/XOTPatternRefinementModel.java:79]: Do you need the default constructor?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/refinementmodels/TopologyFragmentRefinementModelResource.java [hunk 9]
```diff
@@ -127,6 +129,16 @@ public TopologyTemplateResource savePrmMappingTopology(TTopologyTemplate topolog
                     .setArtifactType(QName.valueOf(propertiesMap.get("requiredDeploymentArtifactType")))
                 ));
             }
+            if (relTemplate.getType().getLocalPart().startsWith("BehaviorPatternMapping")) {
```
Comment (lharzenetter) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/refinementmodels/TopologyFragmentRefinementModelResource.java:132]: Maybe also ignore the case...

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java [hunk 10]
```diff
@@ -740,7 +745,6 @@ public static void persist(Object o, RepositoryFileReference ref, MediaType medi
 
     /**
      * @param ref the file to read from
-     * @param repo
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java:743]: Ignore that change or add a short description.

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java [hunk 11]
```diff
@@ -839,9 +843,9 @@ public String getBaseURI() {
 
     /**
      * Derives Winery's Properties Definition from an existing properties definition
-     *  @param ci     the entity type to try to modify the WPDs
+     *
+     * @param ci     the entity type to try to modify the WPDs
      * @param errors the list to add errors to
-     * @param repository
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java:844]: Ignore that change or add a short description.

File: org.eclipse.winery.topologygraph/src/main/java/org/eclipse/winery/topologygraph/matching/patterndetection/ToscaComponentPatternMatcher.java [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        this.refinementModels = refinementModels;
+        this.componentPatternCandidates = componentPatternCandidates;
+        this.componentPatternCandidates.put(prm, new ArrayList<>());
+    }
+
+    @Override
+    public boolean isCompatible(ToscaNode left, ToscaNode right) {
+        return super.isCompatible(left, right)
+            || (!isStayingElement(left)
+            && (shareSupertype(left, right) || componentPatternsCompatible(left, right)));
```
Comment (lharzenetter) [inline org.eclipse.winery.topologygraph/src/main/java/org/eclipse/winery/topologygraph/matching/patterndetection/ToscaComponentPatternMatcher.java:55]: I would add some more parathenses and a different formatting to ease understanding this expression.

File: org.eclipse.winery.topologygraph/src/main/java/org/eclipse/winery/topologygraph/matching/patterndetection/ToscaComponentPatternMatcher.java [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    @Override
+    public boolean isCompatible(ToscaEdge left, ToscaEdge right) {
+        return super.isCompatible(left, right)
+            || (!isStayingElement(left)
+            && (shareSupertype(left, right) || componentPatternsCompatible(left, right)));
```
Comment (lharzenetter) [inline org.eclipse.winery.topologygraph/src/main/java/org/eclipse/winery/topologygraph/matching/patterndetection/ToscaComponentPatternMatcher.java:62]: see above

File: org.eclipse.winery.model.adaptation/src/test/java/org/eclipse/winery/model/adaptation/substitution/patterndetection/BehaviorPatternDetectionTest.java (hunk 2 above)
Comment (awersching) [reply org.eclipse.winery.model.adaptation/src/test/java/org/eclipse/winery/model/adaptation/substitution/patterndetection/BehaviorPatternDetectionTest.java:301]: Yes, I initially tried to use the `MockNamespaceManager` here but it's only accessible in the `topologygraph.matching` package, not in the `adaptation.substitution` package of this test. I could move the `MockNamespaceManager` to src/main of `topologygraph.matching` to make it accessible but I'm not sure if that's a better solution...?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/refinementmodels/TopologyFragmentRefinementModelResource.java (hunk 9 above)
Comment (awersching) [reply org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/refinementmodels/TopologyFragmentRefinementModelResource.java:132]: I'm not exactly sure what you mean, if I ignore this case the Behavior Pattern Mappings won't be saved in the Grafic PRM view. Or should I move the case to the `PatternRefinementModelResource`?

File: org.eclipse.winery.model.adaptation/src/test/java/org/eclipse/winery/model/adaptation/substitution/patterndetection/BehaviorPatternDetectionTest.java (hunk 2 above)
Comment (lharzenetter) [reply org.eclipse.winery.model.adaptation/src/test/java/org/eclipse/winery/model/adaptation/substitution/patterndetection/BehaviorPatternDetectionTest.java:301]: Okay, we use your version here.

**Review by lharzenetter — APPROVED** [review]

**Comment by lharzenetter** [comment]
Looks good! Please create the final PR in [eclipse/winery](https://github.com/eclipse/winery). :)

**Comment by awersching** [comment]
Follow up at https://github.com/eclipse/winery/pull/593

**Pull request closed without merging**